### PR TITLE
New `otelcol.exporter.debug` component

### LINKED ISSUE #5849: Add a new otelcol.exporter.debug component
### Request

We need to add a new `otelcol.exporter.debug` component based on the [debug exporter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/debugexporter). We should also deprecate `otelcol.exporter.debugging` by adding a "warning" message in [its doc](https://grafana.com/docs/agent/latest/flow/reference/components/otelcol.exporter.logging/) that the component is deprecated and may be removed around October 2024.

### Use case

OpenTelemetry Collector's [logging exporter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/loggingexporter) has been [deprecated](https://github.com/open-telemetry/opentelemetry-collector/pull/8378) in favour of the new [debug exporter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/debugexporter).

### PR DESCRIPTION
**Pull request by BarunKGP**
#### PR Description
Added a new `otelcol.exporter.debug` component based on the [otelcol debugexporter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/debugexporter).

#### Which issue(s) this PR fixes

Fixes grafana/alloy#301 

#### Notes to the Reviewer
- Based on same design as other `otelcol.exporter` components, specifically the `otelcol.exporter.logging` component (as debug exporter is replacing logging exporter).
- Uses the same defaults as specified in the [otelcol debugexporter factory](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/debugexporter/factory.go)
- Added `exporterArguments` type which is a wrapper over `Arguments` and maintains an implementation that is consistent with `debugexporter`.
  - Removed `DebugMetrics` field from `Arguments` struct. However, it couldn't be completely removed because the upstream `exporter.Argument` specifies this in its interface.
  - `Arguments.Verbosity` is now a string and can be set to a string (such as `"detailed"`). `exporterArguments` deals with the conversion into `configtelemetry.Level` type
- Added tests to check that component `Config` is created from `Arguments` and error is thrown when an invalid `Verbosity` is passed

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Tests added

### CODE REVIEW

File: component/otelcol/exporter/debug/debug.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			fact := debugexporter.NewFactory()
+			return exporter.New(opts, fact, args.(Arguments), exporter.TypeAll)
+		},
+	})
+}
+
+// Arguments configures the otelcol.exporter.debug component.
+type Arguments struct {
+	Verbosity          configtelemetry.Level `river:"verbosity,attr,optional"`
```
Comment (ptodev) [inline component/otelcol/exporter/debug/debug.go:28]: It'd be good to replace `configtelemetry.Level` with a new `Level` type within the Agent's `component/otelcol/exporter/debug/debug` package. Then it can be translated to the OpenTelemetry type `configtelemetry.Level` using a Convert() function. This has a few benefits:
* We would be able to set `verbosity` via a string (e.g. `verbosity = "detailed"`). I am not sure if we can do it now.
* We get a bit more control of what values can be set in the Agent's config.

File: component/otelcol/exporter/debug/debug.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	SamplingInitial    int                   `river:"sampling_initial,attr,optional"`
+	SamplingThereafter int                   `river:"sampling_thereafter,attr,optional"`
+
+	// DebugMetrics configures component internal metrics. Optional.
+	DebugMetrics otelcol.DebugMetricsArguments `river:"debug_metrics,block,optional"`
```
Comment (ptodev) [inline component/otelcol/exporter/debug/debug.go:33]: I think we don't actually need `debug_metrics` in  `otelcol.exporter.debug`. It's ok to remove it. It is only really useful for exporter/receiver which send/receive data over the network. 

`otelcol.exporter.logging` has it, but I think it was unnecessary that we'd added it there. We can leave it in `otelcol.exporter.logging`, but it would help keep `otelcol.exporter.debug` cleaner and less confusing if we don't have it.

File: go.mod [hunk 3]
```diff
@@ -616,6 +616,7 @@ require (
 	github.com/natefinch/atomic v1.0.1
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/filterprocessor v0.87.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver v0.87.0
+	go.opentelemetry.io/collector/exporter/debugexporter v0.89.0
```
Comment (ptodev) [inline go.mod:619]: There are lots of changes to the go.mod. Updating `go.opentelemetry.io/collector` could impact other components, so we normally update such OTel dependencies in one big batch whenever we need to upgrade to a new version.

I think the reason lots of modules got updated is that `debugexporter` is imported as v0.89.0 instead of v0.87.0. Would you mind resetting the changes to go.mod, and instead just adding this line manually:
```
go.opentelemetry.io/collector/exporter/debugexporter v0.87.0
```
In theory that should be the only necessary change to go.mod.

**Review by ptodev — CHANGES_REQUESTED** [review]
Thank you for your contribution! I left a few comments. Would you mind also adding a couple of things please:
* Unit tests - they need to be in a debug_test.go file, similar to [spanmetrics_test.go](https://github.com/grafana/agent/blob/main/component/otelcol/connector/spanmetrics/spanmetrics_test.go).
* A documentation page - it could be similar to [otelcol.exporter.logging.md](https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/otelcol.exporter.logging.md)

File: go.sum
Comment (ptodev) [inline go.sum:1]: There are still some leftover unnecessary changes in `go.sum`. Would you mind doing a `git reset` on `go.sum`, running `go mod tidy` and rebuilding the Agent please? It shoul help cleanup go.sum.

File: CHANGELOG.md [hunk 4]
```diff
@@ -20,6 +20,8 @@ Main (unreleased)
 
 ### Enhancements
 
+- Added `otelcol.exporter.debug` component based on `go.opentelemetry.io/collector/exporter/debugexporter`. (@BarunKGP)
```
Comment (ptodev) [inline CHANGELOG.md:23]:
```suggestion
- Added a new `otelcol.exporter.debug` component based on OTel Collector's "debug" exporter`. (@BarunKGP)
```

**Review by ptodev — COMMENTED** [review]
Hi, thank you for the update! I don't think it's necessary to have a `exporterArguments` though. It makes adding new arguments more inconvenient, since then we have to add them to both `exporterArguments` and `Arguments` it also makes it less clear how default value are applied, and could make the component a bit more slow due to extra copying.

> I'll add it within the next couple of days, if that's okay?

Yes, of course, thank you!

File: component/otelcol/exporter/debug/debug.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+			return exporter.New(opts, fact, args.(Arguments), exporter.TypeAll)
+		},
+	})
+}
+
+// Arguments configures the otelcol.exporter.debug component.
+type exporterArguments struct {
+	Verbosity          configtelemetry.Level `river:"verbosity, attr, optional"`
+	SamplingInitial    int                   `river:"sampling_initial,attr, optional"`
+	SamplingThereafter int                   `river:"sampling_thereafter,attr, optional"`
```
[the remark this answers is not in the document]
Comment (ptodev) [reply component/otelcol/exporter/debug/debug.go:32]: TBH I'd just remove `exporterArguments`, I don't think it's really necessary. Converting from `Arguments` to the OTel type is ok.

File: go.sum
Comment (BarunKGP) [reply go.sum:1]: Done

File: component/otelcol/exporter/debug/debug.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (BarunKGP) [reply component/otelcol/exporter/debug/debug.go]: Removed `exporterArguments`

File: component/otelcol/exporter/debug/debug_test.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"testing"
+
+	"github.com/grafana/agent/component/otelcol/exporter/debug"
+	"github.com/stretchr/testify/require"
+	otelcomponent "go.opentelemetry.io/collector/component"
+)
+
+func Test(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (BarunKGP) [reply component/otelcol/exporter/debug/debug_test.go:15]: Added tests in table format similar to ones in `spanmetrics_test.go`

**Comment by BarunKGP** [comment]
@ptodev happy new year! Apologies for the delay in making the requested changes, I got caught up in other stuff over the holidays.

I've now addressed all the review comments you left. In summary:
- Reset `go.sum` and reverted to the previous dependency versions
- Added tests to `debug_test.go` in table format and checked that they are passing
- Removed unnecessary `exporterArguments` field 
- Edited changelog for clarity

Please let me know if there are any other changes you'd like me to make. Thanks!

**Comment by ptodev** [comment]
Hi @BarunKGP! Happy new year, and apologies for the late reply! The code looks great, thank you!

The only remaining task is to add a docs page similar to [otelcol.exporter.logging.md](https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/otelcol.exporter.logging.md). Would you be able to add it please?

**Comment by BarunKGP** [comment]
@ptodev I've created the doc, but what about the aliases section as in [otelcol.exporter.logging.md](https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/otelcol.exporter.logging.md)? Should I remove it, or do we have any new aliases to include here?

Also taking a look at the linter issue. I'll push my changes by this weekend.

**Comment by ptodev** [comment]
> @ptodev I've created the doc, but what about the aliases section as in [otelcol.exporter.logging.md](https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/otelcol.exporter.logging.md)? Should I remove it, or do we have any new aliases to include here?

Thank you, the docs look good! It is best to retain the same aliases as `otelcol.exporter.logging`, so the doc is good as it is.

> Also taking a look at the linter issue. I'll push my changes by this weekend.

Thank you! I'll give it a final stamp of approval afterwards. It would be nice to rebase the branch with `main` too.

File: docs/sources/flow/reference/components/otelcol.exporter.debug.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+description: Learn about otelcol.exporter.debug
+labels:
+  stage: experimental
+title: otelcol.exporter.debug
+---
+
+# otelcol.exporter.debug
+
+`otelcol.exporter.debug` accepts telemetry data from other `otelcol` components
+and writes them to the console (stderr). The verbosity of the logs can also be controlled.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.exporter.debug.md:17]:
```suggestion
`otelcol.exporter.debug` accepts telemetry data from other `otelcol` components and writes them to the console (stderr).
You can control the verbosity of the logs.
```

File: docs/sources/flow/reference/components/otelcol.exporter.debug.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+> **NOTE**: `otelcol.exporter.debug` is a wrapper over the upstream
+> OpenTelemetry Collector `debug` exporter. Bug reports or feature requests will
+> be redirected to the upstream repository, if necessary.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.exporter.debug.md:21]:
```suggestion
{{< admonition type="note" >}}
otelcol.exporter.debug` is a wrapper over the upstream OpenTelemetry Collector `debug` exporter.
If necessary, bug reports or feature requests are redirected to the upstream repository.
{{< /admonition >}}
```

File: docs/sources/flow/reference/components/otelcol.exporter.debug.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+Multiple `otelcol.exporter.debug` components can be specified by giving them
+different labels.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.exporter.debug.md:24]:
```suggestion
You can specify multiple `otelcol.exporter.debug` components by giving them different labels.
```

We can break on sentences vs arbitrary (80 column).

File: docs/sources/flow/reference/components/otelcol.exporter.debug.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+configuration.
+
+## Debug information
+
+`otelcol.exporter.debug` does not expose any component-specific debug
+information.
+
+## Example
+
+This example scrapes prometheus unix metrics and writes them to the console:
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/otelcol.exporter.debug.md:67]:
```suggestion
This example scrapes Prometheus UNIX metrics and writes them to the console:
```

**Review by clayton-cornell — COMMENTED** [review]
Some doc input.

**Comment by BarunKGP** [comment]
@clayton-cornell  Thank you for the doc suggestions. I have accepted and incorporated all of them.

@ptodev rebased the branch. Let me know if there's anything else I should take a look at :)

**Comment by ptodev** [comment]
@BarunKGP - The rebase must have gone wrong, because there are lots of changes in the PR now.

**Comment by BarunKGP** [comment]
@ptodev I've reverted this branch to its state prior to the rebase and added the doc changes to it. Upon investigating, it looks like I was incorporating suggestions from these comments onto my origin/main directly, which is why the two branches diverged.

I'm thinking of creating a new branch with all my changes from here, while I force pull `upstream/main` on this one and then rebase my changes on top of it. Unfortunately, the git log will get a bit messy. Do you have any suggestions on how to go about this?

**Comment by ptodev** [comment]
This is the way I normally rebase my feature branches:
```
git fetch
git rebase origin/main
```

If you'd like to clean up your branch in a more step-by-step basis, you can also do something like:
```
# Remove your commits but retain your changes
git reset --soft <comit-before-your-commits>`

git stash

# Make your branch the same as main
# I'm not too sure about this one, but I think it'd work?
git reset --hard main

git stash pop
```

You could also just create a new branch off of main, apply your changes there, and then use `git reset` on your old branch to make it point to the same commit as the new one.

**Comment by BarunKGP** [comment]
I've added all the changes to the branch after rebasing. I've also updated the new components to the revised API for `components`.

What should be the `Stability` for this however? I've left it as `StabilityExperimental` for now.

File: internal/component/otelcol/exporter/debug/debug.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	return verbosity, nil
+}
+
+var _ exporter.Arguments = Arguments{}
+
+// DefaultArguments holds default values for Arguments.
+var DefaultArguments = Arguments{
+	Verbosity:          "basic",
```
Comment (ptodev) [inline internal/component/otelcol/exporter/debug/debug.go:58]:
```suggestion
	Verbosity:          "normal",
```

File: docs/sources/flow/reference/components/otelcol.exporter.debug.md [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+otelcol.exporter.debug "LABEL" { }
+```
+
+## Arguments
+
+`otelcol.exporter.debug` supports the following arguments:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`verbosity`           | `string` | Verbosity of the generated logs. | `"basic"` | no
```
Comment (ptodev) [inline docs/sources/flow/reference/components/otelcol.exporter.debug.md:38]:
```suggestion
`verbosity`           | `string` | Verbosity of the generated logs. | `"normal"` | no
```
The default [upstream](https://github.com/open-telemetry/opentelemetry-collector/tree/v0.86.0/exporter/debugexporter) is "normal", so let's use the same one here.

File: internal/component/otelcol/exporter/debug/debug_test.go [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+		args           string
+		expectedReturn debugexporter.Config
+		errorMsg       string
+	}{
+		{
+			testName: "defaultConfig",
+			args: `
+			verbosity = "basic"
+			sampling_initial = 2
+			sampling_thereafter = 500
```
Comment (ptodev) [inline internal/component/otelcol/exporter/debug/debug_test.go:27]:
```suggestion
```

In order to truly test if the defaults are set correctly, we need to not supply any arguments.

File: go.mod [hunk 14]
```diff
@@ -620,6 +620,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/filterprocessor v0.87.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver v0.87.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver v0.87.0
+	go.opentelemetry.io/collector/exporter/debugexporter v0.87.0
```
Comment (ptodev) [inline go.mod:623]: Are all the changes to the go.mod really necessary? This line is the only one I'd expect us to add.

**Review by ptodev — COMMENTED** [review]
Thank you for the update!

> What should be the `Stability` for this however? I've left it as `StabilityExperimental` for now.

"Experimental" is the right one, yes.

There is a [linter error](https://drone.grafana.net/grafana/agent/17377/1/2):
```
internal/component/otelcol/exporter/debug/debug.go:9: File is not `gofmt`-ed with `-s` (gofmt)
```

I also added a few minor comments. After they are resolved, I think it should be ok for merging.

**Comment by BarunKGP** [comment]
@ptodev thank you for your time in reviewing the PR! 

I have accepted all suggestions and made the following changes addressing them:

- Fixed imports in go.mod -> only added debugexporter as a dependency
- Updated default arguments and test for default arguments
- Fixed linting issues

Please let me know if there are any other areas I need to work on  😃

**Review by ptodev — APPROVED** [review]
Thank you very much! LGTM 🚀

**Comment by clayton-cornell** [comment]
@ptodev This one seems to have been missed... is it still valid? Does anything been to be moved to Alloy?

**Review by clayton-cornell — APPROVED** [review]
Docs are OK as-is

**Comment by clayton-cornell** [comment]
@ptodev There is a linting error (outside the docs) that I can't fix.  Any ideas?

**Comment by ptodev** [comment]
@clayton-cornell I fixed it, no worries. Thank you for the review!
@BarunKGP Thank you for your contribution, and apologies for the delay with the review!

I also opened a [PR](https://github.com/grafana/alloy/pull/1043) to add the component to Alloy.
